flink-runtime/src/test/java/org/apache/flink/runtime/rest/compatibility/CompatibilityCheckResult.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.compatibility;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The (potentially aggregated) result of a compatibility check that may also contain a list of {@link AssertionError}
+ * for each found incompatibility.
+ */
+final class CompatibilityCheckResult {
+
+	private final Compatibility backwardCompatibility;
+	private final int backwardCompatibilityGrade;
+	private final Collection<AssertionError> backwardCompatibilityErrors;
+
+	CompatibilityCheckResult(final Compatibility backwardCompatibility) {
+		this(backwardCompatibility, 1, Collections.emptyList());
+		if (backwardCompatibility == Compatibility.INCOMPATIBLE) {
+			throw new RuntimeException("This constructor must not be used for incompatible results.");
+		}
+	}
+
+	CompatibilityCheckResult(final AssertionError backwardCompatibilityError) {
+		this(Compatibility.INCOMPATIBLE, 0, Collections.singletonList(backwardCompatibilityError));
+	}
+
+	private CompatibilityCheckResult(final Compatibility backwardCompatibility, final int backwardCompatibilityGrade, final Collection<AssertionError> backwardCompatibilityErrors) {
+		this.backwardCompatibility = backwardCompatibility;
+		this.backwardCompatibilityErrors = backwardCompatibilityErrors;
+		this.backwardCompatibilityGrade = backwardCompatibilityGrade;
```

Comment:
nit: `backwardCompatibilityErrors` can be mutable (l. 70) and the getter makes this object potentially mutable